java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -0,0 +1,131 @@
+ * - [DependencyNode.AggregateValue] connects partial access paths to other nodes in the graph.
+ *   These are used to form left-hand-side / right-hand-side relations between handle connections.
+ *
+ * Example:
+ *   ```
+ *   particle HousePets
+ *     input: reads PetData { cats: Number, dogs: Number, household: Text }
+ *     output: writes PetStats {
+ *       ratio: inline Domestication {trained: Number, total: Number},
+ *       family: Text,
+ *       limit: Number
+ *     } = new PetStats {
+ *       ratio: new Domestication {
+ *         trained: input.cats,
+ *         total: input.cats + input.dogs,
+ *       },
+ *       family: input.household,
+ *       limit: 5
+ *     }
+ *   ```
+ *
+ *   This can be translated to:
+ *
+ *   ```
+ *   DependencyNode.AggregateValue(
+ *    "ratio" to DependencyNode.AggregateValue(
+ *      "trained" to DependencyNode.PrimitiveValue("input", "cats"),
+ *      "total" to DependencyNode.DerivedFrom(
+ *        DependencyNode.PrimitiveValue("input", "cats"),
+ *        DependencyNode.PrimitiveValue("input", "dogs")
+ *      )
+ *    ),
+ *    "family" to DependencyNode.PrimitiveValue("input", "household"),
+ *    "limit" to DependencyNode.LITERAL
+ *   )
+ *   ```
+ *
+ *   We can represent this DAG graphically:
+ *
+ *                    (root)
+ *            __________|______
+ *           /      |          \
+ *          /      /        _(ratio)_
+ *        /       /        |         \
+ *   (limit)  (family)  (trained)  (total)
+ *     |        |          |     /   |
+ *     |        |          |   /     |
+ *     x     (household)  (cats)  (dogs)
+ *
+ *   This internal representation, in turn, can be translated into the following claims:
+ *
+ *   ```
+ *   claim output.ratio.trained derives from input.cats
+ *   claim output.ratio.total derives from input.cats and derives from input.dogs
+ *   claim output.family derives from input.household
+ *   ```
+ */
+sealed class DependencyNode {
+
+  /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
+  data class PrimitiveValue(val path: Path = emptyList()) : DependencyNode() {
+    constructor(vararg fields: Identifier) : this(listOf(*fields))
+  }
+
+  /** Represents modification of an access path within a Paxel [Expression]. */
+  data class DerivedFrom(val inputs: Set<DependencyNode> = emptySet()) : DependencyNode() {
+
+    constructor() : this(emptySet())
+
+    constructor(vararg paths: Path) : this(paths.map { PrimitiveValue(it) }.toSet())
+
+    constructor(vararg nodes: DependencyNode) : this(flatten(*nodes))
+
+    companion object {
+      /** Flatten nested sets of [DependencyNode]s.*/
+      private fun flatten(vararg nodes: DependencyNode): Set<DependencyNode> {
+        return nodes.flatMap { node ->
+          when (node) {
+            is PrimitiveValue -> setOf(node)
+            is DerivedFrom -> node.inputs
+            else -> throw IllegalArgumentException(
```

Comment:
This makes me wonder whether we should change the type of `inputs` property to be `Set<Path>` or `Set<PrimitiveNode>`?


Also, see my comment about `AggregateValue` not being a `DependencyNode`.

Reply:
Ah, yes, it was a `Set<Path>` before some recent changes, for this reason.

Reply:
Since I think it makes sense to keep `AggregateValue` (or `PaxelScope`) where it is now, I'm leaning towards factoring this as a set of `Path`s. WDYT?

Reply:
Set<Input>. See my other comment.